src/v/cloud_storage/materialized_resources.cc

```diff
@@ -456,4 +654,68 @@ void materialized_resources::maybe_trim_segment(
     }
 }
 
+class throttled_dl_source : public ss::data_source_impl {
+public:
+    explicit throttled_dl_source(
+      ss::input_stream<char> st,
+      materialized_resources& parent,
+      ss::abort_source& as,
+      ss::gate::holder h)
+      : _src(std::move(st).detach())
+      , _parent(parent)
+      , _as(as)
+      , _holder(std::move(h)) {}
+
+    ss::future<ss::temporary_buffer<char>> get() override {
+        auto h = _gate.hold();
+        return _src.get().then(
+          [this, h = std::move(h)](ss::temporary_buffer<char> buf) mutable {
+              size_t buffer_size = buf.size();
+              return maybe_throttle(buffer_size)
+                .then([b = std::move(buf), h = std::move(h)]() mutable {
+                    return std::move(b);
+                });
+          });
+    }
+
+    ss::future<ss::temporary_buffer<char>> skip(uint64_t n) override {
+        auto holder = _gate.hold();
+        return _src.skip(n).finally(
+          [this, n, h = std::move(holder)] { return maybe_throttle(n); });
+    }
+
+    ss::future<> close() override {
+        _holder.release();
+        return _gate.close().then([this] { return _src.close(); });
+    }
+
+private:
+    ss::future<> maybe_throttle(size_t buffer_size) {
+        auto available = _parent._throughput_limit.available();
+        if (available < buffer_size) {
+            // The throttling will be applied. Register the amount.
+            _parent._read_path_probe.download_throttled(
+              buffer_size - available);
```

Comment:
We should add a metric for the amount of throttled time here. Can be a follow-up though.